sdks/python/apache_beam/yaml/yaml_transform_ut_test.py

```diff
@@ -0,0 +1,301 @@
+          - type: Filter
+            name: FilterOutBigNumbers
+            input: PyMap 
+            keep: "lambda x: x<100"
+          '''
+    return yaml.load(pipeline_yaml, Loader=SafeLineLoader)
+
+  def test_init(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    self.assertEqual(len(scope._transforms_by_uuid), 3)
+    self.assertCountEqual(
+        list(scope._uuid_by_name.keys()),
+        ["PyMap", "Square", "Filter", "FilterOutBigNumbers"])
+
+  def test_get_transform_id_and_output_name(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    transform_id, output = scope.get_transform_id_and_output_name("Square")
+    self.assertEqual(transform_id, spec[0]['__uuid__'])
+    self.assertEqual(output, None)
+
+  def test_get_transform_id_and_output_name_with_dot(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    transform_id, output = \
+      scope.get_transform_id_and_output_name("Square.OutputName")
+    self.assertEqual(transform_id, spec[0]['__uuid__'])
+    self.assertEqual(output, "OutputName")
+
+  def test_get_transform_id_by_uuid(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    transform_id = scope.get_transform_id(spec[0]['__uuid__'])
+    self.assertEqual(transform_id, spec[0]['__uuid__'])
+
+  def test_get_transform_id_by_unique_name(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    transform_id = scope.get_transform_id("Square")
+    self.assertEqual(transform_id, spec[0]['__uuid__'])
+
+  def test_get_transform_id_by_ambiguous_name(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    with self.assertRaisesRegex(ValueError, r'Ambiguous.*PyMap'):
+      scope.get_transform_id(scope.get_transform_id(spec[1]['name']))
+
+  def test_get_transform_id_by_unknown_name(self):
+    spec = self.get_spec()
+    scope = LightweightScope(spec)
+    with self.assertRaisesRegex(ValueError, r'Unknown.*NotExistingTransform'):
+      scope.get_transform_id("NotExistingTransform")
+
+
+def new_pipeline():
+  return beam.Pipeline(
+      options=beam.options.pipeline_options.PipelineOptions(
+          pickle_library='cloudpickle'))
+
+
+class MainTest(unittest.TestCase):
+  def get_scope_by_spec(self, p, spec, inputs=None):
+    if inputs is None:
+      inputs = {}
+    spec = yaml.load(spec, Loader=SafeLineLoader)
+
+    scope = Scope(
+        beam.pvalue.PBegin(p),
+        inputs,
+        spec['transforms'],
+        yaml_provider.standard_providers(), {})
+    return scope, spec
+
+  def test_expand_leaf_transform_with_input(self):
+    with new_pipeline() as p:
+      spec = '''
+          transforms:
+          - type: Create
+            elements: [0]
+          - type: PyMap
+            input: Create
+            fn: "lambda x: x*x"
+          '''
+      scope, spec = self.get_scope_by_spec(p, spec)
+      result = expand_leaf_transform(spec['transforms'][1], scope)
```

Comment:
It's a little unclear what you're trying to test here. I am a bit wary of making tests that are to tied to the particular implementation. Generally I think we want to focus on testing things that (1) are part of the public API and/or (2) invariants or behaviors that if broken would cause issues with the feature.

Reply:
Yeah, I'd remove this test. I think the guiding principle for testing internal things is is "if this changed its behavior, would things break in inscrutable ways?"

Reply:
ok, removing `test_expand_leaf_transform_with_input()` and `test_expand_leaf_transform_without_input()`